server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.identity.schedule;
+
+import java.io.IOException;
+
+import org.opensearch.common.xcontent.XContentParserUtils;
+import org.opensearch.core.xcontent.ToXContentObject;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+
+/**
+ * Class that represents a Scheduled Job Identity
+ */
+public class ScheduledJobIdentityModel implements ToXContentObject {
+    public static final String USER = "user";
+
+    public static final String TOKEN = "token";
+    private final ScheduledJobUserModel user;
+    private final String authToken;
+
+    public ScheduledJobIdentityModel(final ScheduledJobUserModel user, final String authToken) {
+        this.user = user;
+        this.authToken = authToken;
```

Comment:
In the case of plugins, they could pass the `user` in here as its currently stored in many plugins jobs indices. For extensions, they would pass the token since they do not have access to user info on the threadcontext but the token represents the same info.

At REST this data is stored in a system index owned by the security plugin which means that no user, even admin can tamper. It does leave the door open for super admin to modify the identity associated with the job which IMO is desirable. That would allow orphaned references to deleted users or roles to be updated in a central place where its currently confusing being stored in different plugin indices in possibly different ways.

Reply:
Lets go to the high level scenario for a moment because I think it will better inform what data we will have available.

1) As an extension, after I've been initialized, create a job that run as a service account.
2) As an extension, after I've been called by a user, create a job that runs as that user
3) As an extension, after I've been called by a user, create a job that runs as a user other than the user making the call.

For the API call somewhere that creates a job, it takes its existing parameters and it also takes:
- Nothing, and the subject of the current call is used (Covers 1 & 2)
- Principal associated with the job

For the current caller, all the AuthZ information has been loaded, so that works with exactly what you've got planned.

For a passed in principal (Alice wants the job to run as Bob), I think we need a way to generate/hydrate the AuthZ information from a principal - of course checking if this is allowed.

---

However! There is a problem even if we are reusing the current caller, if the AuthZ information expires how do we get it back - which also overlaps with the passed in principal problem space.

Setting aside the details of what is stored within the thread context or how its used.  How can we go from expired authz information -> fresh authz information without the user logging back in?

Reply:
> Setting aside the details of what is stored within the thread context or how its used. How can we go from expired authz information -> fresh authz information without the user logging back in?

In this case its driven by the entry in the centralized scheduled job identity index. If a job is scheduled successfully, an entry is added to the secure and centralized `.opendistro-security-scheduled-job-identity` index. You can think of that entry like a refresh token and its (mostly) tamper proof because its in a system index.

```
.opendistro-security-scheduled-job-identity
{
  "job_id": "<job_id>",
  "job_index": ".opendistro-anomaly-detector-jobs", // This and the field above are a direct reference to the document in .opendistro-anomaly-detector-jobs index 
  "user":
    {
        "name": "anomaly_user",
        "backend_roles":
        [
            "backend_role"
        ],
        "roles":
        [
            "all_access"
        ],
        "custom_attribute_names":
        [],
        "user_requested_tenant": "global"
    }
}
```

This entry drives the creation of fresh authz information on job invocation. For plugins that means injecting roles into the threadcontext, for extensions that means issuing a token with the issue where the entry in the `.opendistro-security-scheduled-job-identity` contains the info needed for token issuance like the username (principal) and authorization info like the mapped roles.

Reply:
The `issueAccessTokenOnBehalfOfUser` method uses the refresh token by reading from `.opendistro-security-scheduled-job-identity` to the authenticating backend to get fresh authz information.  The example above shows the authz information in the new index, is that the refresh token?  How is that used to refresh?

Reply:
Yes treat that entry as a refresh token of sorts - its not a JWT but by having the entry there its assumed that a user has previously given consent to have a job run in the background on their behalf.

Its designed this way to have a migration path from how plugins are scheduling jobs today and makes it pretty clear how to convert a plugin with scheduled jobs (like AD) to an extension. 

Plugins will already store the info above alongside the job information in their jobs index and then inject those `roles` into the threadcontext. By migrating that info from the plugin's job index into a centralized system index that is tamper proof - it is more secure and offloads the work of roles injection from the plugin/extension to the identity system.

It may not be encoded and guaranteed to be tamper-proof like a JWT, but since it is a system index there are guarantees that it cannot be meddled with by anyone with very few exceptions. By storing it like this, it would make it easier to develop APIs are changing the identity or roles that a job runs with.

Reply:
A refresh token is used to create fresh AuthZ information, I don't see how this is possible from the raw information that is placed in the index except for basic auth users.  Please update this thread with data flow diagrams or links to the code that illustrates how this will work.

Reply:
Not sure I am following. The entry in the scheduled job identity index contains all of the information to create new access tokens. If we were to store a refresh token, all of the same info would need to be stored as claims in the refresh token.

```
.opendistro-security-scheduled-job-identity
{
  "job_id": "<job_id>",
  "job_index": ".opendistro-anomaly-detector-jobs", // This and the field above are a direct reference to the document in .opendistro-anomaly-detector-jobs index 
  "user":
    {
        "name": "anomaly_user",
        "backend_roles":
        [
            "backend_role"
        ],
        "roles":
        [
            "all_access"
        ],
        "custom_attribute_names":
        [],
        "user_requested_tenant": "global"
    }
}
```

When creating a new access token, the username (subject) and mapped roles are required to make this backward compatible with how plugins do security for scheduled jobs.